### PR TITLE
TYP: core.arrays.numpy_

### PR DESCRIPTION
**Pull request by simonjayhawkins**

### CODE REVIEW

File: pandas/core/arrays/numpy_.py [hunk 1]
```diff
@@ -35,52 +35,68 @@ class PandasDtype(ExtensionDtype):
 
     Parameters
     ----------
-    dtype : numpy.dtype
+    dtype : object
+        Object to be converted to a NumPy data type object.
```
Comment (jbrockmendel) [inline pandas/core/arrays/numpy_.py]: i see how this is more accurate than just "numpy.dtype", but can we be more specific here?

File: pandas/core/arrays/numpy_.py [hunk 2]
```diff
@@ -35,52 +35,68 @@ class PandasDtype(ExtensionDtype):
...
+    --------
+    numpy.dtype
     """
 
     _metadata = ("_dtype",)
 
-    def __init__(self, dtype):
-        dtype = np.dtype(dtype)
-        self._dtype = dtype
-        self._type = dtype.type
```
Comment (jbrockmendel) [inline pandas/core/arrays/numpy_.py]: @jorisvandenbossche does turning this into properties have perf implications we need to worry about

File: pandas/core/arrays/numpy_.py (hunk 1 above)
Comment (simonjayhawkins) [reply pandas/core/arrays/numpy_.py]: it's the types accepted by np.dtype which is more permissive than just another np.dtype. Object is  straight out of the numpy docs.

do you want to restrict this as before, or establish the actual types accepted by numpy?

File: pandas/core/arrays/numpy_.py (hunk 2 above)
Comment (simonjayhawkins) [reply pandas/core/arrays/numpy_.py]: this is for consistency with name that was changed in #31037

File: pandas/core/arrays/numpy_.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/arrays/numpy_.py]: if its straight out of the numpy docs, then sounds fine

File: pandas/core/arrays/numpy_.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/core/arrays/numpy_.py]: OK thanks.  That was the perf affect I had in mind, but pre-coffee i had it backwards.

File: pandas/core/arrays/numpy_.py [hunk 3]
```diff
@@ -35,52 +35,68 @@ class PandasDtype(ExtensionDtype):
...
+        """
         return self._dtype
 
     @property
-    def name(self):
+    def name(self) -> str:
+        """
+        A bit-width name for this data-type.
+
+        Un-sized flexible data-type objects do not have this attribute.
```
Comment (jorisvandenbossche) [inline pandas/core/arrays/numpy_.py]: I think you can leave this out

File: pandas/core/arrays/numpy_.py [hunk 4]
```diff
@@ -91,15 +107,23 @@ def construct_array_type(cls):
...
-    def itemsize(self):
-        """The element size of this data-type object."""
+    def itemsize(self) -> int:
+        """
+        The element size of this data-type object.
+        """
         return self._dtype.itemsize
 
 
+PandasArrayT = TypeVar("PandasArrayT", bound="PandasArray")
```
Comment (jorisvandenbossche) [inline pandas/core/arrays/numpy_.py]: Why is this needed?

File: pandas/core/arrays/numpy_.py (hunk 3 above)
Comment (simonjayhawkins) [reply pandas/core/arrays/numpy_.py]: no problem. will remove. can we wrap all numpy dtypes?

File: pandas/core/arrays/numpy_.py (hunk 4 above)
Comment (simonjayhawkins) [reply pandas/core/arrays/numpy_.py]: L166 is type self, so needed for constructor signature to ensure values is type self.

Comment (jorisvandenbossche) [reply pandas/core/arrays/numpy_.py]: Don't we use the name as a string in other places?

Comment (simonjayhawkins) [reply pandas/core/arrays/numpy_.py]: removed for now.

Comment (jorisvandenbossche) [reply pandas/core/arrays/numpy_.py]: To be clear, I didn't want to imply you necessarily needed to remove it, rather I want to understand it (eg what is the difference with using the class name as a string as we do elsewhere)

To me, this typing is adding a lot of black magic to the codebase, and I want to ensure it stays as understandable as possible for people (including me) that are not so familiar with typing.

Comment (simonjayhawkins) [reply pandas/core/arrays/numpy_.py]: an example for the case where the return type is from a base class method

```
class Base:
    def foo(self) -> "Base":
        return type(self)()


class Derived(Base):
    x = 0

    def bar(self) -> int:
        return self.foo().x


print(Derived().bar())
```

This is valid python...

```
$ python pandas/test.py
0
```
however, mypy will report errors...
```
$ mypy pandas/test.py
pandas\test.py:10: error: "Base" has no attribute "x"
Found 1 error in 1 file (checked 1 source file)
```

whereas using a typevar...
```python
from typing import TypeVar

_T = TypeVar("_T")


class Base:
    def foo(self: _T) -> _T:
        return type(self)()


class Derived(Base):
    x = 0

    def bar(self) -> int:
        return self.foo().x


print(Derived().bar())
```
gives
```
$ mypy pandas/test.py
Success: no issues found in 1 source file
```

so typing the base class needs to use typevars to keep mypy green when we add types to the subclasses. I have reverted, but will probably need to add back when the subclasses get merged.

I've not included an example for the init case, but similar principal

Comment (jorisvandenbossche) [reply pandas/core/arrays/numpy_.py]: OK, that explanation helps (but if we start adding it, it needs to be explained in our contributing guidelines)

But: `PandasArray` is not a base class, it is not further subclassed I think. So why is it needed here?

Comment (jorisvandenbossche) [reply pandas/core/arrays/numpy_.py]: Ah, actually, StringArray now subclasses PandasArray, so forget the last sentence of what I just said ;)

**Review by WillAyd — APPROVED** [review]

**Comment by WillAyd** [comment]
Thanks!

**Comment by jorisvandenbossche** [comment]
@simonjayhawkins a follow-up documenting this practice somewhere would be welcome!
